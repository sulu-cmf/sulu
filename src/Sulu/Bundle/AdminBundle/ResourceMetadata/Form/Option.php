<?php

/*
 * This file is part of Sulu.
 *
 * (c) MASSIVE ART WebServices GmbH
 *
 * This source file is subject to the MIT license that is bundled
 * with this source code in the file LICENSE.
 */

namespace Sulu\Bundle\AdminBundle\ResourceMetadata\Form;

class Option
{
    /**
     * @var null|string
     */
    protected $name;

    /**
     * @var string|int|Option[]
     */
    protected $value;

    /**
     * @var ?string
     */
    protected $title;

    /**
     * @var ?string
     */
    protected $placeholder;

    /**
     * @var ?string
     */
    protected $infoText;

    /**
     * @return null|string|int|float
     */
    public function getName()
    {
        return $this->name;
    }

    /**
     * @param null|string|int|float $name
     */
    public function setName($name = null): void
    {
        $this->name = $name;
    }

    /**
     * @return int|string|Option[]
     */
    public function getValue()
    {
        return $this->value;
    }

    /**
     * @param int|string|Option[] $value
     */
    public function setValue($value): void
    {
        $this->value = $value;
    }

    public function addValueOption(self $option): void
    {
        if (!is_array($this->value)) {
            $this->value = [];
        }

        $this->value[] = $option;
    }

    public function getTitle(): ?string
    {
        return $this->title;
    }

    public function setTitle(string $title = null): void
    {
        $this->title = $title;
    }

    public function getPlaceholder(): ?string
    {
        return $this->placeholder;
    }

    public function setPlaceholder(string $placeholder = null): void
    {
        $this->placeholder = $placeholder;
    }

    public function getInfotext(): ?string
    {
        return $this->infoText;
    }

    public function setInfotext(string $infoText = null): void
    {
        $this->infoText = $infoText;
    }
}
