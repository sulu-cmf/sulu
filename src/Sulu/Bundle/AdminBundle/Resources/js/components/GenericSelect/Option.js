// @flow
import React from 'react';
import type {ElementRef} from 'react';
import classNames from 'classnames';
import {afterElementsRendered} from '../../services/DOM';
import Icon from '../Icon';
import Checkbox from '../Checkbox';
import type {OptionSelectedVisualization} from './types';
import optionStyles from './option.scss';

type Props = {
    selected: boolean,
    disabled: boolean,
    focus: boolean,
    value: string,
    children: string,
    onClick?: (value: string) => void,
    optionRef?: (optionNode: ElementRef<'li'>, selected: boolean) => void,
    selectedVisualization: OptionSelectedVisualization,
};

const SELECTED_ICON = 'check';

export default class Option extends React.PureComponent<Props> {
    static defaultProps = {
        disabled: false,
        selected: false,
        focus: false,
        selectedVisualization: 'icon',
        value: '',
    };

    item: ElementRef<'li'>;

    handleButtonClick = () => {
        if (this.props.onClick) {
            this.props.onClick(this.props.value);
        }
    };

    setItem = (item: ElementRef<'li'>) => {
        const {
            selected,
            optionRef,
        } = this.props;

        if (optionRef) {
            optionRef(item, selected);
        }
    };

    setButton = (button: ElementRef<'button'>) => {
        if (!button) {
            return;
        }

        if (this.props.focus) {
            afterElementsRendered(() => {
                button.focus();
            });
        }
    };

    renderSelectedVisualization() {
        if (this.props.selectedVisualization === 'icon') {
            return this.props.selected ? <Icon className={optionStyles.icon} name={SELECTED_ICON} /> : null;
        }

        return (
            <Checkbox
                onChange={this.handleButtonClick}
                className={optionStyles.input}
                checked={this.props.selected}
            />
        );
    }

    render() {
        const {
            selected,
            children,
            disabled,
            selectedVisualization,
        } = this.props;
        const optionClass = classNames(
            optionStyles.option,
            optionStyles[selectedVisualization],
            {
                [optionStyles.selected]: selected,
            },
        );

        return (
            <li ref={this.setItem}>
                <button
                    className={optionClass}
                    ref={this.setButton}
                    onClick={this.handleButtonClick}
                    disabled={disabled}
                >
                    {this.renderSelectedVisualization()}
                    {children}
                </button>
            </li>
        );
    }
}
