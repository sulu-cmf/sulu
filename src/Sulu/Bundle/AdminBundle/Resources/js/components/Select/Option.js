// @flow
import React from 'react';
import type {ElementRef} from 'react';
import classNames from 'classnames';
import {afterElementsRendered} from '../../services/DOM';
import Icon from '../Icon';
import Checkbox from '../Checkbox';
import type {OptionSelectedVisualization} from './types';
import optionStyles from './option.scss';

type Props = {
    anchorWidth: number,
    selected: boolean,
    disabled: boolean,
    focus: boolean,
    value: string | number,
    children: string,
    onClick?: (value: string | number) => void,
    optionRef?: (optionNode: ElementRef<'li'>, selected: boolean) => void,
    selectedVisualization: OptionSelectedVisualization,
};

const ANCHOR_WIDTH_DIFFERENCE = 10;

export default class Option extends React.PureComponent<Props> {
    static defaultProps = {
        anchorWidth: 0,
        disabled: false,
        focus: false,
        selected: false,
        selectedVisualization: 'icon',
        value: '',
    };

    item: ElementRef<'li'>;

    handleButtonClick = () => {
        if (this.props.onClick) {
            this.props.onClick(this.props.value);
        }
    };

    setItemRef = (item: ?ElementRef<'li'>) => {
        const {
            selected,
            optionRef,
        } = this.props;

        if (optionRef && item) {
            optionRef(item, selected);
        }
    };

    setButtonRef = (button: ?ElementRef<'button'>) => {
        if (this.props.focus) {
            afterElementsRendered(() => {
                if (!button) {
                    return;
                }

                button.focus();
            });
        }
    };

    renderSelectedVisualization() {
        if (this.props.selectedVisualization === 'icon') {
            return this.props.selected ? <Icon className={optionStyles.icon} name={'su-check'} /> : null;
        }

        return (
            <Checkbox
                checked={this.props.selected}
                className={optionStyles.input}
                onChange={this.handleButtonClick}
            />
        );
    }

    render() {
        const {
            anchorWidth,
            selected,
            children,
            disabled,
            selectedVisualization,
        } = this.props;
        const optionClass = classNames(
            optionStyles.option,
            optionStyles[selectedVisualization],
            {
                [optionStyles.selected]: selected,
            }
        );

        return (
            <li ref={this.setItemRef}>
                <button
                    className={optionClass}
                    disabled={disabled}
                    onClick={this.handleButtonClick}
                    ref={this.setButtonRef}
                    style={{minWidth: anchorWidth + ANCHOR_WIDTH_DIFFERENCE}}
                >
                    {this.renderSelectedVisualization()}
                    {children}
                </button>
            </li>
        );
    }
}
