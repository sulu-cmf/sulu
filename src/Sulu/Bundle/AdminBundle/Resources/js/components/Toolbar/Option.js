// @flow
import classNames from 'classnames';
import React from 'react';
import Icon from '../Icon';
import type {Skin} from './types';
import optionStyles from './option.scss';

const ICON_CHECKMARK = 'su-check';

type Props = {
    label: string | number,
    value: Object,
    onClick: (value: Object) => void,
    size?: string,
    skin?: Skin,
    selected?: boolean,
    disabled?: boolean,
};

export default class Option extends React.PureComponent<Props> {
    handleOnClick = () => {
        const {onClick} = this.props;

        onClick(this.props.value);
    };

    render() {
        const {
            skin,
            size,
            label,
            selected,
            disabled,
        } = this.props;
        const optionClass = classNames(
            optionStyles.option,
            optionStyles[skin],
            {
                [optionStyles[size]]: size,
                [optionStyles.isSelected]: selected,
            }
        );

        return (
            <li className={optionClass}>
                <button
                    disabled={disabled}
                    onClick={this.handleOnClick}
                >
                    {selected &&
                        <Icon name={ICON_CHECKMARK} className={optionStyles.selectedIcon} />
                    }
                    {label}
                </button>
            </li>
        );
    }
}
