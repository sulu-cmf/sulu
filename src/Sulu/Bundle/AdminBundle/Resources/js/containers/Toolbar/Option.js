// @flow
import classNames from 'classnames';
import React from 'react';
import Icon from '../../components/Icon';
import optionStyles from './option.scss';

const ICON_CHECKMARK = 'check';

type Props = {
    label: string | number,
    value: Object,
    onClick: (value: Object) => void,
    size?: string,
    selected?: boolean,
    disabled?: boolean,
};

export default class Option extends React.PureComponent<Props> {
    handleOnClick = () => {
        const {onClick} = this.props;

        onClick(this.props.value);
    };

    render() {
        const {
            size,
            label,
            selected,
            disabled,
        } = this.props;
        const optionClasses = classNames({
            [optionStyles.option]: true,
            [optionStyles[size]]: size,
            [optionStyles.isSelected]: selected,
        });

        return (
            <li className={optionClasses}>
                <button
                    disabled={disabled}
                    onClick={this.handleOnClick}>
                    {selected &&
                        <Icon name={ICON_CHECKMARK} className={optionStyles.selectedIcon} />
                    }
                    {label}
                </button>
            </li>
        );
    }
}
