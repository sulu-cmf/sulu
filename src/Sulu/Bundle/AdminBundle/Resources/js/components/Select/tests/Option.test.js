/* eslint-disable flowtype/require-valid-file-annotation */
import {render, mount, shallow} from 'enzyme';
import React from 'react';
import Option from '../Option';

jest.mock('../../../services/DOM/afterElementsRendered');

test('The component should render', () => {
    const option = render(<Option value="my-option">My option</Option>);
    expect(option).toMatchSnapshot();
});

test('The component should render in selected state', () => {
    const option = render(<Option value="my-option" selected={true}>My option</Option>);
    expect(option).toMatchSnapshot();
});

test('The component should render with checkbox', () => {
    const option = render(<Option value="my-option" selectedVisualization="checkbox">My option</Option>);
    expect(option).toMatchSnapshot();
});

test('The component should render in disabled state', () => {
    const option = render(<Option value="my-option" disabled={true}>My option</Option>);
    expect(option).toMatchSnapshot();
});

test('A click on the component should fire the callback', () => {
    const clickSpy = jest.fn();
    const option = shallow(<Option onClick={clickSpy}>My option</Option>);
    option.find('button').simulate('click');
    expect(clickSpy).toBeCalled();
});

test('The component should be focused when the corresponding property was set', (done) => {
    expect(document.activeElement.tagName).toBe('BODY');
    mount(<Option focus={true}>My option</Option>);
    setTimeout(() => {
        expect(document.activeElement.tagName).toBe('BUTTON');
        done();
    });
});
