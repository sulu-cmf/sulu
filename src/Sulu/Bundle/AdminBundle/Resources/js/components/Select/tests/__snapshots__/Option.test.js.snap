// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`The component should render 1`] = `
<li>
  <button
    class="option icon"
    style="min-width:10px"
  >
    My option
  </button>
</li>
`;

exports[`The component should render in disabled state 1`] = `
<li>
  <button
    class="option icon"
    disabled=""
    style="min-width:10px"
  >
    My option
  </button>
</li>
`;

exports[`The component should render in selected state 1`] = `
<li>
  <button
    class="option icon selected"
    style="min-width:10px"
  >
    <span
      aria-label="su-check"
      class="su-check icon"
    />
    My option
  </button>
</li>
`;

exports[`The component should render with checkbox 1`] = `
<li>
  <button
    class="option checkbox"
    style="min-width:10px"
  >
    <label
      class="label"
    >
      <span
        class="switch checkbox dark input"
      >
        <input
          type="checkbox"
        />
        <span />
      </span>
    </label>
    My option
  </button>
</li>
`;
